site/_posts/2024-11-27-fish-4.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: "fish-shell 4.0b and the Rust port"
+date: 2024-11-09
+categories: general
+---
+
+fish is a smart and user-friendly command line shell (like bash or zsh) with clever features that work out of the box, with nothing new to learn or configure.
+
+fish-shell is approaching its 4.0 milestone, the first release since porting from C++ to Rust. Today we are announcing an open beta, inviting all users to try out the upcoming 4.0 release.
+
+### **Try the fish-shell 4.0 beta at [fishshell.com](https://fishshell.com)**. View the [release notes here](https://fishshell.com/docs/current/relnotes.html).
+
+If you are an existing fish user, or are curious to give it a try, this is a great opportunity to improve your command line experience while contributing to fish. Please report any issues you encounter by filing a [GitHub Issue](http://github.com/fish-shell/fish-shell/issues), or on our [Matrix channel](https://matrix.to/#/#fish-shell:matrix.org), which is also a great place to ask questions. Newbie questions welcome!
+
+The remainder of this post will discuss the mechanics of the port.
+
+### By the numbers
+
+The port was proposed in February 2023. January 2024 marked the elimination of C++ from the fish-shell codebase; this was then followed by a period of testing, fixes, and preparation for the release.
+
+Prior to the port, fish-shell had over 47k lines of code in C++ source files, and nearly 8k lines in C++ header files. Today, fish has over 75k lines of Rust. This increase is explained by Rust's explicit `use` constructs compared to C++'s umbrella headers, `rustfmt` vs `clang-format` differences, the need to re-implement `printf`, and some additional features added in the interim.
+
+### Why port
+
+fish-shell was originally written in C, in 2005. It was ported to C++ in 2011, for improved memory management, and C++ served fish-shell well for a decade.
+
+But eventually, problems arose. Modern C++ is challenging to learn and makes it difficult to attract new developers. fish-shell maintains compatibility with older operating systems, limiting us to C++11 with no clear upgrade path. Lastly, fish-shell has a long-term goal of enabling fully multithreaded execution, and prototypes of this in C++ proved unpromising. For example, it was too easy to accidentally share objects across threads, with only post-hoc tools like Thread Sanitizer to prevent it.
```

Comment:
Something along the lines of:

We try to make it easy for everyone to get the newest version of fish even on the long term support systems that are commonly used as servers. So we both build our own packages for these and try to make it buildable with the system's compiler, because upgrading the C++-compiler is a pain. To that end we limited ourselves to C++11 (though later C++ versions did not offer a lot to help with our pain points).

Rust makes this easier by offering new releases via rustup.

?

Reply:
I'm not convinced that it's rustup that has made life easy here - I had to wrangle new versions of Rust onto the Linux package builders, which I have done in the past with GCC as well. In some ways the fact that everyone expects an up-to-date version has encouraged more work on packaging it for older distributions, which is work I have relied on.

Reply:
Sure, but it has made life easier *for users* to build on their systems.

I've run an old NetBSD release and just installed rustup and built fish. Had we expected a C++ compiler of similar age that would have been a massive pain.

Reply:
Great point, that would have been a non-starter.

Reply:
One upgrade path would be the one that the distributions intend to be used.
The decision on whether new fish should be available on old platforms needn't necessarily be made by fish authors.